File: lib/src/main/java/vserver/route/VariableSubPath.java
package vserver.route;

import vserver.SubPath;
import vserver.RoutingContext;

public class VariableSubPath implements SubPath {
    private final SubPath next;
    private final String variable;

    public VariableSubPath(SubPath next, String variable) {
        this.next = next;
        this.variable = variable;
    }

    @Override
    public SubPath next() {
        return next;
    }

    @Override
    public boolean match(String route) {
        return true;
    }

    @Override
    public boolean currentSame(SubPath r) {
        return r instanceof VariableSubPath && ((VariableSubPath) r).variable.equals(variable);
    }

    @Override
    public void fill(RoutingContext ctx, String route) {
        ctx.putParam(variable, route);
    }
}

File: lib/src/main/kotlin/io/vproxy/lib/http/route/VariableSubPath.kt
package io.vproxy.lib.http.route

import vproxy.lib.http.RoutingContext
import java.net.URLDecoder
import java.nio.charset.StandardCharsets

class VariableSubPath(private val next: SubPath?, private val variable: String) : SubPath {
  override fun next(): SubPath? {
    return next
  }

  override fun match(route: String): Boolean {
    return true
  }

  override fun currentSame(r: SubPath): Boolean {
    return r is VariableSubPath && r.variable == variable
  }

  override fun fill(ctx: RoutingContext, route: String) {
    val res = try {
      URLDecoder.decode(route, StandardCharsets.UTF_8)
    } catch (ignore: RuntimeException) {
      route
    }
    ctx.putParam(variable, res)
  }

  override fun toString(): String {
    return "/:$variable"
  }
}
